articles/containerization-and-why-to-use-it.md

```diff
@@ -0,0 +1,58 @@
+---
+authorName: Florian Besser
+authorGithubUsername: florian-besser
+issue: 76
+title: Containerization and why to use it
+---
+# {{page.title}}
+
+You are a good software engineer, but you dread releases as they are known as a time of pain. Your project matches one or more of the following:
+* Deployments are done manually or via scripts that execute on some remote host.
+* Scaling the app is hard as new machines or VMs take weeks to arrive & setup.
+* Your application dependencies differ in version or feature set between environments.
+* Developers have difficulty reproducing production issues on their local machines.
+* Application dependencies are installed manually, developers spend days setting up their own machine.
+
+This article will explain how containerization can help solve each of these issues.
+
+## Cumbersome deployments and dirty hosts
+
+Running scripts on a host seems OK in the beginning, but over time your host will become "dirty". Some versions of your script were only run in QA, and production has leftover artifacts from previous deployments. These leftovers confuse maintainers and reduce confidence in your team's ability to deploy software to production systems.
+
+With containerization, you immediately get rid of leftovers your software may cause. By **throwing away the old container** after deploying a new version you have already cleaned up after your application.
+
+A tangent about volumes/host mounts: Sometimes your app writes files outside of its container. In these cases the files still need to be cleaned up, but usually an app has a well-defined output directory that rarely changes.
+
+Volumes allow you to use containerization with databases quite nicely: The data is safe from crashes but up- and downgrading a DB is just a question of stopping the old container and booting a new one.
```

Comment:
In diesem Abschnitt verstehe ich jetzt den Punkt nicht. Vermutlich liegt es daran, dass ich nicht weiss, was du mit "Volumes" meinst. Ich vermute: Volumes ist das was benötigt wir um "Sometimes your app writes files outside of its container."
Ist die DB in einem Container? Oder eben ein Volume?

Reply:
Volumes ist wo du Daten reinschreibst die nicht ephemeral sind, d.h. einen Container Crash / Neustart überleben müssen.

Die DB ist also im Container, die Daten liegen im Volume. Um die DB zu upgraden wird der Container removed und ein neuer Container gestartet. Beide Container verwenden dasselbe Volume insofern bleiben die Daten erhalten.